### PR TITLE
wip: add socks5 support

### PR DESCRIPTION
**Pull request by thomasf**
Work in progress for #9 

## basic configuration

I'm working towards the configuration style below. The `ports:` field is based
on code for a docker-compose.yml `ports:"` parser from some docker package..

```yml

  myservice:
    method: http
    http:
      ports:
        - 8100-8120                  # myservice:8100 to myservice:8120
        - 192.168.1.1:8100-8120:8300 # myservice:8300 to myservice:8320

  myservice2:
    method: http
    http:
      port: 8100                     # myservice2:1 or myservice2:8100
```

## easy/safe unique scrape host names

The config below would create the same scrape target instance names
`myservice:1` and `myservice:2` for multiple actual hosts which isn't great and
can lead to accidental mix ups in Prometheus queries. 

```yml
 - job_name: 'myservice_host1'
    scrape_interval: 1s
    static_configs:
      - targets: ['myservice:1', 'myservice:2']
    scheme: https
    proxy_url: "socks5://host1:3000"
    metrics_path: /proxy
    tls_config:
        ...

 - job_name: 'myservice_host2'
    scrape_interval: 1s
    static_configs:
      - targets: ['myservice:1', 'myservice:2']
    scheme: https
    proxy_url: "socks5://host2:3000"
    metrics_path: /proxy
    tls_config:
        ...

```

While it is probably possible to avoid instance name collisions using
relabeling inside Prometheus I think it's best to be able to ensure instance
name uniqueness at the node level.

To solve this issue an optional `-socks.prefix` flag is introduced to
`exporter_exporter` which enforces a FQN prefix to the virtual host names.

If an `exporter_exporter` instance is started with `-socks.prefix
cool.host1` the scrape target would then look like this:

```yml
 - job_name: 'myservice_host1'
    scrape_interval: 1s
    static_configs:
      - targets: ['cool.host1.myservice:1', 'cool.host1.myservice:2']
    scheme: https
    proxy_url: "socks5://host1:3000"
    metrics_path: /proxy
    tls_config:
        ...

```

To lessen the requirements of manual node configuration An additional boolean
flag `-socks.auto-prefix` will try to set the prefix using `os.Hostname()`
returns unless it returns an empty string or error in which case start up will
be aborted by an fatal exit.

## command line options
right now it looks like this (the descriptions will be revised later):
```
  -socks.listen-address string
    	The address to listen on for accessing the metrics end points via made up internal host names.
  -socks.tls
    	Use TLS with the socks proxy (default true)
  -socks.prefix
        ...
  -socks.auto-prefix (default False)
        ...        
```

## (QUESTION) can not use socks method + TLS + non-TLS at the same time

Because `-socks.tls` and `-socks.listen-address` are combined the socks method
can not be enabled for both tls and non tls access simultaneously. I don't see
it useful to have both working at the same time but we have to change some
things if it should be supported.

## (QUESTION) should it be allowed to configure the same service multiple times?

I'm a bit skeptical of this feature since it complicates the rules for what a
valid configuration can be while not adding strong practical benefits.

**This question can be deferred since it can be added later if needed**

```yml
  myservice:
    method: http
    http:
      ports:
        - 8100-8120                  # myservice:8100 to myservice:8120
        - 192.168.1.1:8100-8120:8300 # myservice:8300 to myservice:8320

  myservice: # invalid config since port 8100 already was defined above
    method: http
    http:
      port: 8100

  myservice: # valid config since port 8400 isn't alloceted for myservice
    method: http
    http:
      ports:
      - "8101:8400"
```

For the directory based config `myservice.app1.yml` and `myservice.app2.yml`
would work for allowing multiple instances.

## (QUESTION) exposing multi port services for non socks methods

**This question can be deferred since it can be added later if needed**

I'm not sure if if is meaningful to implement the multiple ports per service
scheme for the regular http and https serving options via URL parameters. The
corresponding prometheus.yml would become very complicated. Could be useful if
one wants to use a web browser to easily view the exporters but that isn't easy
when using client certificates login anyway so maybe not a big win?

### CODE REVIEW

**Comment by thomasf** [comment]
I'm working towards the configuration style below. The `ports:` field is based
on code for a docker-compose.yml `ports:"` field.


```yml
  blackbox:
    method: http
    http:
       port: 9115
       path: probe


  myservice:  ## same service can be specified multiple times as long as the exposed ports are unque
    method: http
    http:
      ports:
        - 8100-8120                  # myservice:8100 to myservice:8120
        - 192.168.1.1:8100-8120:8300 # myservice:8300 to myservice:8320

  myservice2:
    method: http
    http:
      port: 8100                     # myservice2:0 or myservice2:8100
```


I'm not sure if if is meaningful to implement the multiple ports per service scheme for the regular http and https serving options.

**Comment by thomasf** [comment]
@tcolgate  could you comment on the user facing behavior stuff I wrote about in this issue description.. I would like to have most of it decided before I'm committing my time towards the underlying solution

**Comment by tcolgate** [comment]
`(QUESTION) should it be allowed to configure the same service multiple times?`
It's questionable if multi identical key sections in yaml is valid. I've seen app drop earlier sections, and I've also got a recollection that the latest go-yaml considers this an actual error. I wouldn't bother.

**Comment by tcolgate** [comment]
```  
myservice2:
    method: http
    http:
      port: 8100                     # myservice2:1 or myservice2:8100```
The ambiguity of have name:ordinal and name:port makes me a little uneasy.whilst overlap is incredibly unlikely, it could be confusing to users.

**Comment by tcolgate** [comment]
re: mixing tls and non-tls socks, I'm not too fussed. you could cofigure two seperate listeners I suppose, but I think if someone is already enforcing socks and tls, they are probably fully invested.

**Comment by tcolgate** [comment]
regarding `-socks.prefix cool.host1`, I think it may be better to just support general addition of a label to the returned metrics. Each exporter could then be configured to add its own "proxy_instance" label then

File: expexp.yaml [hunk 1]
```diff
@@ -25,3 +25,15 @@ modules:
...
+    method: http
+    http:
+      ports:
+        - 8100-8120                  # myservice:8100 to myservice:8120
+        - 192.168.1.1:8100-8120:8300 # myservice:8300 to myservice:8320
+
+  myservice2:
+    method: http
+    http:
+      port: 8100                     # myservice:0 or myservice:8100
```
Comment (tcolgate) [inline expexp.yaml]: I'm not sure the ":0" option here adds much.

File: main.go [hunk 2]
```diff
@@ -22,18 +22,20 @@ import (
...
 	"net/http"
 	"os"
 	"path"
 	"path/filepath"
 	"strings"
 	"time"
 
-	"golang.org/x/sync/errgroup"
-
+	"github.com/QubitProducts/exporter_exporter/pkg/socks5"
```
Comment (tcolgate) [inline main.go]: Please just vendor/ this and give appropriate attribute in a NOTICE flile

File: main.go [hunk 3]
```diff
@@ -22,18 +22,20 @@ import (
...
 	"github.com/golang/glog"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/valyala/fasthttp/fasthttputil"
```
Comment (tcolgate) [inline main.go]: I'd rather just copy in the inmemory listener code here and give attribution in a NOTICES file. We don't need anything else from fasthttp, and it's a pretty simple type.

**Comment by tcolgate** [comment]
@thomasf do you have an actual production use case for the port ranges?

File: main.go (hunk 2 above)
Comment (thomasf) [reply main.go]: It needed some changes.

Comment (thomasf) [reply main.go]: https://github.com/QubitProducts/exporter_exporter/pull/10/commits/d09ed2205dd357ecb88ffa1bcb8aa0dd5c4ae8aa

File: main.go (hunk 3 above)
Comment (thomasf) [reply main.go]: yeah.. this was just a poc implementation to see that it actually worked.. Don't look to close at how the pull request diffs.. the description text is the important parts for now.

**Comment by thomasf** [comment]
Yes, python application clusters where each process has a separate prometheus exporter running on it;s own thread.

File: main.go (hunk 2 above)
Comment (thomasf) [reply main.go]: That change is maybe not the real solution, just something to get it working for the POC.

**Comment by tcolgate** [comment]
thanks, generally in favour. It does feel like maybe it's got a few extra features that aren't specific to socks implementation.

**Comment by thomasf** [comment]
while this has worked in practise for a very long time in production i'm closing this issue.. we have changed to just using a wildcard domain for each host so that each prometheus exporter gets it'd own https subdomain instead.

**Comment by thomasf** [comment]
if anyone is interested we monitor like this now

```yml
  - job_name: "minio-foobar-staging"
    scheme: https
    metrics_path: '/minio/prometheus/metrics'
    static_configs:
      - targets: ['s.foobar.staging.example.com']
        labels:
          context: 'foobar.staging'

  - job_name: "netdata-foobar-staging"
    basic_auth:
      username: prometheus
      password: 'somepassword'
    scheme: https
    metrics_path: '/api/v1/allmetrics'
    params:
      format: ['prometheus']
    static_configs:
      - targets: ['netdata.monitor.foobar.staging.example.com']
        labels:
          context: 'foobar.staging'

  - job_name: "traefik-foobar-staging"
    basic_auth:
      username: prometheus
      password: 'somepassword'
    scheme: https
    static_configs:
      - targets: ['traefik.monitor.foobar.staging.example.com']
        labels:
          context: 'foobar.staging'

  - job_name: "django-foobar-staging"
    basic_auth:
      username: prometheus
      password: 'somepassword'
    scheme: https
    static_configs:
      - targets: ['app1.monitor.foobar.staging.example.com', 'app2.monitor.foobar.staging.example.com', 'crawl1.monitor.foobar.staging.example.com','notify1.monitor.foobar.staging.example.com']
        labels:
          context: 'foobar.staging'
```

**Pull request closed without merging**